src/main/java/org/terasology/metalrenegades/combat/system/EnemySpawnSystem.java

```diff
@@ -0,0 +1,209 @@
+
+                if (enemyGracePeriodComponent.cyclesLeft >= 0) {
+                    character.removeComponent(EnemyGracePeriodComponent.class);
+                } else {
+                    character.saveComponent(enemyGracePeriodComponent);
+                }
+            }
+
+            if (!character.hasComponent(SettlementRefComponent.class)) {
+                spawnEnemyOnCharacter(character);
+            }
+        }
+
+        enemyQueue.removeIf(enemy -> {
+            LocationComponent locComp = enemy.getComponent(LocationComponent.class);
+            Vector3f enemyLoc = locComp.getWorldPosition();
+            if (settlementEntityManager.checkOutsideAllSettlements(new Vector2i(enemyLoc.getX(), enemyLoc.getZ()))) {
+                return false;
+            }
+            enemy.destroy();
+            return true;
+        });
+
+        while (enemyQueue.size() > MAX_ENEMIES) {
+            enemyQueue.remove().destroy();
+        }
+    }
+
+    @ReceiveEvent
+    public void onDawnEvent(OnDawnEvent event, EntityRef entityRef) {
+        while (!enemyQueue.isEmpty()) {
+            enemyQueue.remove().destroy();
+        }
+    }
+
+    @ReceiveEvent
+    public void onCharacterRespawn(OnPlayerRespawnedEvent event, EntityRef entity) {
+        entity.saveComponent(new EnemyGracePeriodComponent(5));
+    }
+
+    private void spawnEnemyOnCharacter(EntityRef character) {
+        LocationComponent locationComponent = character.getComponent(LocationComponent.class);
+
+        List<Vector3i> spawnPositions = findSpawnPositions(new Vector3i(locationComponent.getWorldPosition()));
+        Optional<Vector3i> potentialPos = spawnPositions.stream().filter(p -> random.nextFloat() > 0.999).findFirst();
+
+        if (potentialPos.isPresent()) {
+            spawnOnPosition(potentialPos.get());
+        }
+    }
+
+    private List<Vector3i> findSpawnPositions(Vector3i centrePos) {
+        Vector3i worldPos = new Vector3i(centrePos);
+        List<Vector3i> foundPositions = Lists.newArrayList();
+        Vector3i blockPos = new Vector3i();
+        for (int y = -10; y < 10; y++) {
+            for (int z = -30; z < 30; z++) {
+                for (int x = -30; x < 30; x++) {
+                    blockPos.set(x + worldPos.x, y + worldPos.y, z + worldPos.z);
+                    if (isValidSpawnPosition.apply(blockPos)) {
+                        foundPositions.add(new Vector3i(blockPos));
+                    }
+                }
+            }
+        }
+        return foundPositions;
+    }
+
+    private boolean isValidSpawnPosition(Vector3i pos) {
+        if (!settlementEntityManager.checkOutsideAllSettlements(new Vector2i(pos.x, pos.z))) {
+            return false;
+        }
+
+        Vector3i below = new Vector3i(pos.x, pos.y - 1, pos.z);
+        Block blockBelow = worldProvider.getBlock(below);
+        if (!blockBelow.equals(sand)) {
+            return false;
+        }
```

Comment:
Maybe change the condition to `if(blockBelow.isPenetrable())` ?. Because I have seen some cases when mineral ores have formed at the top layer. Also makes it generic, if we decide to add different biome features such as oasis or chasms, that might use blocks that are not sand.

Reply:
These conditions were taken from the deer spawning system, where the deer shouldn't spawn on cactus or inside buildings. However, for enemies it probably doesn't matter as much, so it's been swapped out 👍